src/mococrw/p11_wrap.h

```diff
@@ -0,0 +1,223 @@
+/*
+ * #%L
+ * %%
+ * Copyright (C) 2022 BMW Car IT GmbH
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+#pragma once
+
+#include "p11_lib.h"
+#include "util.h"
+
+namespace mococrw
+{
+namespace p11
+{
+/**
+ * Template to wrap OpenSSLs "_free" functions
+ * into a functor so that a std::unique_ptr
+ * can use them.
+ */
+template <class R, class P, R(Func)(P *)>
+struct P11Deleter
+{
+    template <class T>
+    void operator()(T *ptr) const noexcept
+    {
+        if (ptr) {
+            Func(ptr);
+        }
+    }
+};
+
+/**
+ * An Exception for LibP11 errors.
+ *
+ * This exception is thrown by all methods when an OpenSSL error occurs.
+ */
+class P11Exception final : public std::exception
+{
+public:
+    template <class StringType>
+    explicit P11Exception(StringType &&message) : _message{std::forward<StringType>(message)}
+    {
+    }
+
+    /**
+     * Generate an exception with the defalt LibP11 error-string
+     * as message.
+     *
+     */
+    P11Exception() : _message{generateP11ErrorString()} {}
+
+    const char *what() const noexcept override { return _message.c_str(); }
+
+private:
+    static std::string generateP11ErrorString();
+    std::string _message;
+};
+
+/*
+ * Wrap all the pointer-types returned by LibP11.
+ */
+
+using P11_PKCS11_CTX_Ptr =
+        std::unique_ptr<PKCS11_CTX, P11Deleter<void, PKCS11_CTX, lib::LibP11::P11_PKCS11_CTX_free>>;
+
+/*
+ * The memory referred to by these pointer types live out of scope of their pointers.
+ * Therefore, simply wrap the C pointer, without use of smart pointer (this is a bit ugly).
+ */
+using P11_PKCS11_SLOT_PTR = PKCS11_SLOT *;
```

Comment:
It was more of a means to avoid using directly LibP11 definitions at upper levels of the MoCOCrW's code-base.

What do you mean by:

> They're used inconsistently in the rest of this file anyways.